src/security/core/tests/authentication.c

```diff
@@ -95,57 +141,276 @@ static void authentication_init(bool different_ca, const char * trusted_ca_dir,
+  CU_ASSERT_EQUAL_FATAL (read_fail, exp_read_fail);
 }
 
-static void authentication_fini(bool delete_pp)
+static void authentication_fini(bool delete_pp1, bool delete_pp2)
 {
-  if (delete_pp)
-  {
+  if (delete_pp1)
     CU_ASSERT_EQUAL_FATAL (dds_delete (g_participant1), DDS_RETCODE_OK);
+  if (delete_pp2)
     CU_ASSERT_EQUAL_FATAL (dds_delete (g_participant2), DDS_RETCODE_OK);
-  }
   CU_ASSERT_EQUAL_FATAL (dds_delete (g_domain1), DDS_RETCODE_OK);
   CU_ASSERT_EQUAL_FATAL (dds_delete (g_domain2), DDS_RETCODE_OK);
 }
 
-CU_Test(ddssec_authentication, different_ca)
+#define FM_CA "error: unable to get local issuer certificate"
+#define FM_INVK "Failed to finalize verify context"
+CU_TheoryDataPoints(ddssec_authentication, id_ca_certs) = {
+    CU_DataPoints(const char *,
+    /*                         */"valid ID1-ID1",
+    /*                          |     */"non-trusted remote CA ID1(CA1)-ID2(CA2)",
+    /*                          |      |     */"valid ID1-ID3",
+    /*                          |      |      |     */"invalid ca ID1(CA1)-ID1(CA2)",
+    /*                          |      |      |      |     */"invalid remote key ID3(K1)"),
+    CU_DataPoints(const char *, ID1,   ID2,   ID3,   ID1,   ID3),      /* Identity for domain 2 */
+    CU_DataPoints(const char *, ID1K,  ID2K,  ID3K,  ID1K,  ID1K),     /* Private key for domain 2 identity */
+    CU_DataPoints(const char *, CA1,   CA2,   CA1,   CA2,   CA1),      /* CA for domain 2 identity */
+    CU_DataPoints(bool,         false, false, false, true,  false),    /* expect validate local failed for domain 2 */
+    CU_DataPoints(const char *, NULL,  NULL,  NULL,  FM_CA, NULL),     /* expected error message for validate local failed */
+    CU_DataPoints(bool,         false, false, false, false, true),     /* expect handshake request failed */
+    CU_DataPoints(const char *, NULL,  NULL,  NULL,  NULL,  FM_INVK),  /* expected error message for handshake request failed */
+    CU_DataPoints(bool,         false, true,  false, true,  false),    /* expect handshake reply failed */
+    CU_DataPoints(const char *, NULL,  FM_CA, NULL,  FM_CA, NULL)      /* expected error message for handshake reply failed */
+};
+
+static void validate_hs(struct Handshake *hs, bool exp_fail_hs_req, const char * fail_hs_req_msg, bool exp_fail_hs_reply, const char * fail_hs_reply_msg)
 {
-  authentication_init (true, NULL, false);
-  validate_handshake (DDS_DOMAINID1, true, NULL, true, "error: unable to get local issuer certificate");
-  validate_handshake (DDS_DOMAINID2, true, NULL, true, "error: unable to get local issuer certificate");
-  authentication_fini (true);
+  DDS_Security_ValidationResult_t exp_result = hs->node_type == HSN_REQUESTER ? DDS_SECURITY_VALIDATION_OK_FINAL_MESSAGE : DDS_SECURITY_VALIDATION_OK;
+  if (hs->node_type == HSN_REQUESTER && exp_fail_hs_req)
+  {
+    CU_ASSERT_EQUAL_FATAL (hs->finalResult, exp_fail_hs_req ? DDS_SECURITY_VALIDATION_FAILED : exp_result);
```

Comment:
Execution never gets here if ¬exp_fail_hs_req — I do think one would want to check the result in the case of an expected successful handshake, and that therefore this needs a bit of rework.